### PR TITLE
Remove old components when deploying on a circle (Override deployment) 

### PR DESCRIPTION
**Pull request by cpgo**
By default all deployments performed on a circle will always remove the old components.
If everything work as expected we can remove the undeployment call that Moove performs on each deployment since its always handled by Butler.

### CODE REVIEW

File: butler/src/app/v2/core/integrations/octopipe/request-builder.ts [hunk 1]
```diff
@@ -123,19 +123,32 @@ export class OctopipeRequestBuilder {
...
+
+    }
+    return componentsToBeRemoved(deployment, activeComponents).map(component => {
+      return {
+        componentName: component.name,
+        helmRepositoryConfig: this.getHelmRepositoryConfig(component, deployment.cdConfiguration),
+        helmConfig: this.getHelmConfig(component, deployment.circleId),
+        rollbackIfFailed: false
+
       }
```
Comment (lucasbfernandes) [inline butler/src/app/v2/core/integrations/octopipe/request-builder.ts:150]: Maybe separate these two implementations in different methods?

You could change the deployment object to:

```
{
...
unusedDeployments: deployment.defaultCircle ? defaultUnusedVersions() : circleUnusedVersions()
}
```

File: butler/src/app/v2/core/integrations/spinnaker/pipeline-builder.ts [hunk 2]
```diff
@@ -179,15 +179,23 @@ export class SpinnakerPipelineBuilder {
...
+    }
+
     const evalStageId: number = DeploymentTemplateUtils.getProxyEvalStageId(deployment.components)
-    deployment.components.forEach(component => {
-      const unusedComponent: Component | undefined = DeploymentUtils.getUnusedComponent(activeComponents, component, deployment.circleId)
-      if (unusedComponent) {
-        stages.push(getDeleteUnusedStage(unusedComponent, deployment.cdConfiguration, this.currentStageId++, evalStageId, deployment.circleId))
-      }
+    return componentsToBeRemoved(deployment, activeComponents).map(component => {
+      return getDeleteUnusedStage(component, deployment.cdConfiguration, this.currentStageId++, evalStageId, deployment.circleId)
```
Comment (lucasbfernandes) [inline butler/src/app/v2/core/integrations/spinnaker/pipeline-builder.ts:197]: Maybe separate these two implementations in different methods?

File: butler/src/app/v2/core/integrations/utils/deployment.utils.ts [hunk 3]
```diff
@@ -50,3 +50,36 @@ const DeploymentUtils = {
...
+  return !deploymentComponents?.some(dc => removedConditions(dc, activeComponent))
+}
+
+const updatedComponents = (deploymentComponents: DeploymentComponent[] | undefined, activeComponent: Component) => {
+  return deploymentComponents?.some(dc => updatedConditions(dc, activeComponent))
+}
+
+const removedConditions = (deploymentComponent: DeploymentComponent, activeComponent: Component): boolean => {
+  return isSameName(deploymentComponent, activeComponent)
+}
```
Comment (lucasbfernandes) [inline butler/src/app/v2/core/integrations/utils/deployment.utils.ts:73]: Why have this method if you can put `isSameName(...)` inline on the removedComponents method?

File: butler/src/app/v2/core/integrations/utils/deployment.utils.ts [hunk 4]
```diff
@@ -50,3 +50,36 @@ const DeploymentUtils = {
...
+
+const updatedConditions = (deploymentComponent: DeploymentComponent, activeComponent: Component): boolean => {
+  return isSameNameAndDifferenteVersion(deploymentComponent, activeComponent)
+}
```
Comment (lucasbfernandes) [inline butler/src/app/v2/core/integrations/utils/deployment.utils.ts:77]: Why have this method if you can put `isSameNameAndDifferentVersion(...)` inline on the removedComponents method?

File: butler/src/app/v2/core/integrations/utils/deployment.utils.ts [hunk 5]
```diff
@@ -50,3 +50,36 @@ const DeploymentUtils = {
...
+
+const isSameNameAndDifferenteVersion = (deploymentComponent: DeploymentComponent, activeComponent: Component): boolean => {
```
Comment (lucasbfernandes) [inline butler/src/app/v2/core/integrations/utils/deployment.utils.ts:79]: Typo here: Different not Differente

**Review by lucasbfernandes — CHANGES_REQUESTED** [review]

File: moove/application/src/main/kotlin/io/charlescd/moove/application/deployment/impl/CreateDeploymentInteractorImpl.kt [hunk 6]
```diff
@@ -47,7 +47,6 @@ open class CreateDeploymentInteractorImpl @Inject constructor(
 
         if (build.canBeDeployed()) {
             val deployment = createDeployment(request, workspaceId)
-            undeployActiveDeploymentFromCircleIfExists(deployment, workspaceId)
```
Comment (thallesfreitaszup) [inline moove/application/src/main/kotlin/io/charlescd/moove/application/deployment/impl/CreateDeploymentInteractorImpl.kt:50]: Agora que a chamada do undeploy foi removida do moove acho que seria interessante no callback do deploy atualizar os status dos deployments anteriores para NOT_DEPLOYED, algo que tinhamos no callback do undeploy. Como isso foi removido todo deploy ficaria com status DEPLOYED no moove.

**Review by thallesfreitaszup — CHANGES_REQUESTED** [review]

**Review by thallesfreitaszup — APPROVED** [review]

File: moove/application/src/main/kotlin/io/charlescd/moove/application/deployment/impl/CreateDeploymentInteractorImpl.kt (hunk 6 above)
Comment (lucasbfernandes) [reply moove/application/src/main/kotlin/io/charlescd/moove/application/deployment/impl/CreateDeploymentInteractorImpl.kt:50]: Verdade... Temos que ver essa questão dos status na exibiçao da tela. Sem o undeploy via moove o status vai ficar DEPLOYED até sumir (acho)

**Review by thallesfreitaszup — APPROVED** [review]

**Review by lucasbfernandes — APPROVED** [review]

**Pull request merged**